File: examples/acquisitions_handling.py
'''
Upper-level interface demo that illustrates how MR data can be interfaced 
from python.
'''

import argparse
import os
try:
    import pylab
    HAVE_PYLAB = True
except:
    HAVE_PYLAB = False
import sys

BUILD_PATH = os.environ.get('BUILD_PATH') + '/xGadgetron'
SRC_PATH = os.environ.get('SRC_PATH') + '/xGadgetron/pGadgetron'

sys.path.append(BUILD_PATH)
sys.path.append(SRC_PATH)

from pGadgetron import *

parser = argparse.ArgumentParser(description = \
'''
Upper-level interface demo that illustrates how MR data can be interfaced 
from python.
''')
parser.add_argument\
('filename', nargs='?', default = 'simulated_MR_2D_cartesian.h5', \
 help = 'raw data file name (default: simulated_MR_2D_cartesian.h5)')
args = parser.parse_args()                                 

def main():

    # acquisitions will be read from an HDF file args.filename
    input_data = AcquisitionData(args.filename)

    # Get number of acquisitions:
    # the raw k-space data is a list of different 1D 
    # acquisitions (readouts) of different data type (e.g. noise correlation 
    # data, navigator data, image data,...).
    # The number of all aquisitions is
    na = input_data.number_of_acquisitions()
    # The number of image data acquisitions is
    ni = input_data.number_of_acquisitions('image')
    print('acquisitions: total %d, image data %d' % (na, ni))

    # sort data acquisition
    input_data.sort()

    where = range(254, 258)
    # inspect some acquisitions flags
    flags = input_data.get_info('flags')
    if flags[0] & IMAGE_DATA_MASK:
        print('first acquisition is image data')
    else:
        # should see this if input data file is test_2D_2x.h5
        print('first acquisition is not image data')
        
    # display flags for readout number 254 to 257    
    print('Flags'),
    print(flags[where])
    
    # inspect some kspace_encode_step_1 counters
    es1 = input_data.get_info('encode_step_1')
    print('Ky/PE - encoding'),
    print(es1[where])
    
    # inspect some slice counters
    slices = input_data.get_info('slice')
    print('Slices'),
    print(slices[where])
    
    # inspect some repetition counters
    print('Repetitions'),
    reps = input_data.get_info('repetition')
    print(reps[where])

    # copy raw data into python array and determine its size
    # in the case of the provided dataset 'simulated_MR_2D_cartesian.h5' the 
    # size is 2x256 phase encoding, 8 receiver coils and points 512 readout 
    # points (frequency encoding dimension)
    input_array = input_data.as_array()
    input_shape = input_array.shape
    print('input data dimensions: %dx%dx%d' % input_shape)

    # pre-process acquisitions
    print('---\n pre-processing acquisitions...')
    processed_data = PreprocessAcquisitions(input_data)

    # copy processed acquisitions into an array and determine its size
    # by removing the oversampling factor of 2 along the readout direction, the
    # number of readout samples was halfed
    processed_array = processed_data.as_array()
    processed_shape = processed_array.shape
    print('processed data dimensions: %dx%dx%d' % processed_shape)

try:
    main()
    print('done')

except error as err:
    # display error information
    print('??? %s' % err.value)

File: examples/acquisitions_handling.m
% Upper-level interface demo, illustrates acquisitions pre-processing and plotting.

set_up_mr

try
    % acquisitions will be read from this HDF file
    input_data = AcquisitionData('simulated_MR_2D_cartesian.h5');
    
    na = input_data.number();
    fprintf('%d acquisitions found\n', na)

    fprintf('sorting acquisitions...\n')
    input_data.sort()

    [ns, ny, nc] = input_data.slice_dimensions();

    [ns, nc, na] = input_data.dimensions('image');
    
    % pre-process acquisition data
    fprintf('processing acquisitions...\n')
    processed_data = preprocess_acquisitions(input_data);
    processed_data.sort()
    
    flags = input_data.get_info('flags');
    es1 = input_data.get_info('encode_step_1');
    slice = input_data.get_info('slice');
    repetition = input_data.get_info('repetition');
    
    while true
        num = input('enter acquisition number: ');
        if num < 1 || num > na
            break
        end
        a = input_data.acquisition(num);
        fprintf('samples: %d\n', a.number_of_samples())
        fprintf('flags: %d %d\n', a.flags(), flags(num))
        fprintf('encode step 1: %d %d\n', a.idx_kspace_encode_step_1(), es1(num))
        fprintf('slice: %d %d\n', a.idx_slice(), slice(num))
        fprintf('repetition: %d %d\n', a.idx_repetition(), repetition(num))
    end
    
    iarr0 = input_data.as_array();
    parr0 = processed_data.as_array();
    
    is = ns/2;
    ic = nc/2;
    ia = na/2;
    disp(iarr0(is, ic, ia))
    iarr0(is, ic, ia) = iarr0(is, ic, ia)*10;
    
    input_data.fill(iarr0);
    processed_data.fill(parr0);

    iarr = input_data.as_array();
    parr = processed_data.as_array();

    disp(iarr(is, ic, ia))

    iarr = abs(iarr);
    parr = abs(parr);
    iarr = iarr/max(max(max(iarr)));
    parr = parr/max(max(max(parr)));

    nz = idivide(na, ny);
    fprintf...
        ('Enter z-coordinate of the slice to view the acquired data for it\n')
    fprintf...
        ('(a value outside the range [1 : %d] will stop this loop)\n', nz)
    while (true)
        z = int32(input('slice: '));
        if z < 1 || z > nz
            break
        end
        fprintf('Enter coil number to view the acquired data for it\n')
        fprintf...
            ('(a value outside the range [1 : %d] will stop this loop)\n', nc)
        while (true)
            c = input('coil: ');
            if c < 1 || c > nc
                break
            end
            figure(c)
            data = squeeze(iarr(:, c, ny*(z - 1) + 1 : ny*z));
            imshow(data)
            title('oversampled data')
            cc = double(c + nc);
            figure(cc)
            data = squeeze(parr(:, c, ny*(z - 1) + 1 : ny*z));
            imshow(data)
            title('de-oversampled data')
        end
    end
    
catch err
    % display error information
    fprintf('%s\n', err.message)
    fprintf('error id is %s\n', err.identifier)
end
